### PR TITLE
Seeding

### PR DESCRIPTION
**Pull request by anty-filidor**

### CODE REVIEW

File: network_diffusion/mln/functions.py [hunk 1]
```diff
@@ -26,6 +26,31 @@
...
+
+    for l_name in net.layers:
+        grap: nx.Graph = net.layers[l_name]
+        l_graph[l_name] = nx.betweenness_centrality(grap)
+    for actor in net.get_actors():
+        for l_name in actor.layers:
+            bet[actor] = l_graph[l_name][actor.actor_id]
+    return bet
+
+def closenes(net: MultilayerNetwork) -> Dict[MLNetworkActor, float]:
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:43]: typo

File: network_diffusion/mln/functions.py [hunk 2]
```diff
@@ -26,6 +26,31 @@
 from network_diffusion.mln.mlnetwork import MultilayerNetwork
 
 
+def betweennes(net: MultilayerNetwork) -> Dict[MLNetworkActor, int]:
+    """Return value of centrality for actors from the network"""
+    bet: Dict[MLNetworkActor, int] = {}
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:31]: these two functions return a dict keyed by MLNactor with values of betweenness/closeness in a rendomly chosen layer! This approach will not work

File: network_diffusion/seeding/betweenness_selector.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    def actorwise(self, net: MultilayerNetwork) -> List[MLNetworkActor]:
+        """Get ranking for actors using Betweenness Centrality metric."""
+        betweennes_centrality_values: Dict[MLNetworkActor, float] = {}
+        ranking_list: List[MLNetworkActor] = []
+
+        for actor, a_betweennes in betweennes(net=net).items():
+            #a betweennes's value is added to dict
+            betweennes_centrality_values[actor] = a_betweennes
+            #sorted by value of dictonaries
+        for dc_val, val in sorted(betweennes_centrality_values.items(), reverse=True, key=lambda x: x[1]):
```
Comment (anty-filidor) [inline network_diffusion/seeding/betweenness_selector.py:41]: does it work?

File: network_diffusion/mln/functions.py [hunk 4]
```diff
@@ -26,6 +26,52 @@
...
+        b_graph[l_name] = nx.betweenness_centrality(grap)
+    for actor in net.get_actors():
+        for l_name in actor.layers:
+            bet.append(b_graph[l_name][actor.actor_id])
+        bet_mean[actor] = sum(bet)/len(bet)
+        bet.clear()
+    return bet_mean
+
+def closenes(net: MultilayerNetwork) -> Dict[MLNetworkActor, float]:
+    """Return value of centrality for actors from the network"""
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:46]: centrality?

File: network_diffusion/mln/functions.py [hunk 5]
```diff
@@ -26,6 +26,52 @@
...
+        c_graph[l_name] = nx.closeness_centrality(grap)  # values of closeness for one layer
+    for actor in net.get_actors():
+        for l_name in actor.layers:
+            close.append(c_graph[l_name][actor.actor_id])  # saving information to the dict
+        close_mean[actor] = sum(close)/len(close)
+        close.clear()
+    return close_mean
+
+def katz(net: MultilayerNetwork) -> Dict[MLNetworkActor, float]:
+    """Return value of centrality for actors from the network"""
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:61]: centrality?

File: network_diffusion/seeding/katz_selector.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            f"{BOLD_UNDERLINE}\nseed selection method\n{THIN_UNDERLINE}\n"
+            f"\tkatz centrality choice\n{BOLD_UNDERLINE}\n"
+        )
+
+    @staticmethod
+    def _calculate_ranking_list(graph: nx.Graph) -> List[Any]:
+        """Create nodewise ranking."""
+        raise NotImplementedError(
+            "Nodewise ranking list cannot be computed for this class!"
+        )
```
Comment (anty-filidor) [inline network_diffusion/seeding/katz_selector.py:29]: i think we can get rid of this redundancy

File: network_diffusion/seeding/closeness_selector.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+            f"{BOLD_UNDERLINE}\nseed selection method\n{THIN_UNDERLINE}\n"
+            f"\tcloseness centrality choice\n{BOLD_UNDERLINE}\n"
+        )
+
+    @staticmethod
+    def _calculate_ranking_list(graph: nx.Graph) -> List[Any]:
+        """Create nodewise ranking."""
+        raise NotImplementedError(
+            "Nodewise ranking list cannot be computed for this class!"
+        )
```
Comment (anty-filidor) [inline network_diffusion/seeding/closeness_selector.py:29]: i think we can get rid of this redundancy

File: network_diffusion/mln/functions.py [hunk 8]
```diff
@@ -26,6 +26,63 @@
 from network_diffusion.mln.mlnetwork import MultilayerNetwork
 
 
+def betweenness(net: MultilayerNetwork) -> Dict[MLNetworkActor, float]:
+    """Return value of betweennes centrality for actors from the network."""
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:30]: Return value of mean betweennes centrality of actors across layers.

File: network_diffusion/mln/functions.py [hunk 9]
```diff
@@ -26,6 +26,63 @@
...
+    for actor in net.get_actors():
+        for l_name in actor.layers:
+            bet.append(b_graph[l_name][actor.actor_id])
+        bet_mean[actor] = sum(bet) / len(bet)
+        bet.clear()
+    return bet_mean
+
+
+def closeness(net: MultilayerNetwork) -> Dict[MLNetworkActor, float]:
+    """Return value of closeness centrality for actors from the network."""
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:47]: Return value of mean closeness centrality of actors across layers.

File: network_diffusion/mln/functions.py [hunk 10]
```diff
@@ -26,6 +26,63 @@
...
+            close.append(
+                c_graph[l_name][actor.actor_id]
+            )  # saving information to the dict
+        close_mean[actor] = sum(close) / len(close)
+        close.clear()
+    return close_mean
+
+
+def katz(net: MultilayerNetwork) -> Dict[MLNetworkActor, float]:
+    """Return value of Katz centrality for actors from the network."""
```
Comment (anty-filidor) [inline network_diffusion/mln/functions.py:67]: Return value of mean Katz centrality of actors across layers.

File: network_diffusion/seeding/katz_selector.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
+"""A definition of the seed selector based on closeness centrality."""
```
Comment (anty-filidor) [inline network_diffusion/seeding/katz_selector.py:1]: Katz centrality

File: network_diffusion/tests/test_experiment_logger.py [hunk 12]
```diff
@@ -254,10 +254,8 @@ def test__convert_logs(self):
...
         for phenomena, stat in logger._global_stats_converted.items():
-            self.assertTrue(
-                exp_logs_converted[phenomena].equals(stat),
-                f"Logs for phenomena {phenomena} incorrect. Expected value"
-                f" {exp_logs_converted[phenomena]}, got {stat}",
+            pd.testing.assert_frame_equal(
+                exp_logs_converted[phenomena], stat, check_dtype=False
             )
 
     def test__add_log(self):
```
Comment (anty-filidor) [inline network_diffusion/tests/test_experiment_logger.py:261]: In addition to your changes, please bump the version of the library (I'd recommend `__version__ = "0.10.0"`). Moreover, please update a facade of the selectors: (`/network_diffusion/seeding/__init__.py`) to facilitate usage of the seed selection methods.

**Review by anty-filidor — COMMENTED** [review]
Please review my comments

File: network_diffusion/seeding/cbim.py [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+    @dataclass
+    class ListsEws:
+        """Valuables for lists in ews function."""
+
+        actor_list: List[MLNetworkActor] = field(default_factory=list)
+        sort_list: List[MLNetworkActor] = field(default_factory=list)
+        ews_actor: Dict[MLNetworkActor, int] = field(default_factory=dict)
+        degree_actor: Dict[MLNetworkActor, int] = field(default_factory=dict)
+        seed_size: float = 0
+
```
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py:55]: when i use [] it shows error

File: network_diffusion/seeding/cbim.py [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+from network_diffusion.seeding.degreecentrality_selector import (
+    DegreeCentralitySelector,
+)
+from network_diffusion.utils import BOLD_UNDERLINE, THIN_UNDERLINE
+
+
+class CBIMselector(BaseSeedSelector):
+    """CBIM seed selector."""
+
+    def __init__(self, threshold: float, seed_size: int = None) -> None:
```
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py:22]: corrected

File: network_diffusion/seeding/cbim.py [hunk 15]
```diff
@@ -0,0 +1,571 @@
...
+            net, communities
+        )  # merge communities
+
+        for l_name in net.layers:
+            if self.seed_default:  # select seed_size again for calculate ews
+                self.ListsEws.seed_size = len(net.layers[l_name])
+            else:
+                self.ListsEws.seed_size = self.seed_proc * len(
+                    net.layers[l_name]
+                )
```
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py:155]: i check it

File: network_diffusion/seeding/cbim.py [hunk 16]
```diff
@@ -0,0 +1,571 @@
...
+
+    def __str__(self) -> str:
+        """Return seed method's description."""
+        return (
+            f"{BOLD_UNDERLINE}\nseed selection method\n{THIN_UNDERLINE}\n"
+            f"\tCBIM choice\n{BOLD_UNDERLINE}\n"
+        )
+
+    @dataclass
+    class ListsEws:
```
[the remark this answers is not in the document]
Comment (anty-filidor) [reply network_diffusion/seeding/cbim.py:47]: indeed

File: network_diffusion/seeding/cbim.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (anty-filidor) [reply network_diffusion/seeding/cbim.py:55]: ok, so let's keep that

File: network_diffusion/seeding/cbim.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py]: I corrected my code

File: network_diffusion/seeding/cbim.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py]: It's already been outside

File: network_diffusion/seeding/cbim.py [hunk 17]
```diff
@@ -0,0 +1,571 @@
...
+        super().__init__()
+        assert (
+            seed_size is None or 0 <= seed_size <= 100
+        ), f"incorrect seed_size value: {seed_size}!"
+        if seed_size is None:
+            self.seed_default = True  # seed_default say if seed_size is None
+        else:
+            self.seed_default = False
+            self.seed_proc = (
+                seed_size / 100
```
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py]: I removed it i leaved float

File: network_diffusion/seeding/cbim.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (damian4060) [reply network_diffusion/seeding/cbim.py]: After i changed class to outside your case it's working

**Comment by anty-filidor** [comment]
renamed branch to feature--new-ss-methods

**Pull request closed without merging**